keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,282 @@
+
+Legend:
+- Operators and operator-tasks are marked with double letters e.g. 'AA' or `BB`
+- other tasks are marked with single letters e.g. 'D'
+- direct children of an operator are the 'deploy' plan steps e.g. for 'AA' deploy steps are 'BB', 'CC' and 'D'
+```
+
+In the first step we build a dependency graph. A set of all graph vertices (which are task-operators) `S` is defined as `S = {AA, BB, CC, EE, GG}`. A transitive relationship `R` between the vertices is defined as `(a, b) ∈ S` meaning _`a` needs `b` deployed first_. The transitive relationship for the above example is: `R = { (AA,BB), (AA,CC), (BB,EE), (BB,GG) }`. The resulting topological order `O` is therefor `O = (EE, GG, BB, CC, AA)` which has no cycles.
+
+The instance controller (IC) then starts with the execution of the top-level `deploy` plan of the operator `AA`. The first task is the `BB` operator-task. When executing it, IC creates the `Instance-BB` resource and ends current reconciliation. Next, IC notices new `Instance-BB` resource, starts new reconciliation, and executes the  `deploy` plan of the operator `BB` which then creates `Instance-EE` resource. This way we are basically performing the depth-first search for the dependency graph, executing each vertex in the right order e.g. `EE` has to be healthy before `BB` deploy plan can continue with the next step `F`.
+
+We would additionally add the higher-level `Instance` reference (e.g. `AA`) to the `ownerReferences` list of its direct children `Instance`s (e.g. `BB` and `CC`). This would help with determining which `Instance` belongs to which operator and additionally help us with [operator uninstalling](#uninstalling).
+
+The status of the execution can be seen as usual as part of the `Instance.Status`. We could additionally forward the status of a dependency `Instance` to the top-level `Instance.Status` to simplify the overview.
+
+Note that in the above example if e.g. `EE` and `CC` task-operators reference the same operator package they must use distinct instance names `spec.instanceName` so that two separate `Instance`s are deployed. Otherwise, the dependency graph will have a cycle.
+
+##### Dependencies Parametrization
+
+We want to encourage operator composition by providing a way of operator encapsulation. In other words, operator users should not be allowed to arbitrarily modify the parameters of embedded operator instances. The higher-level operator should define all parameters that its **direct** dependency  operators need. Let's demonstrate this on an example of a simple operator `AA` that has operator `BB` as a dependency. 
+
+```yaml
+AA
+└── BB
+```
+Operator `BB` has a required and empty parameter `PASSWORD`. To provide a way for the `AA` operator user to set the password we extend the operator-task with a new field `parameterFile`:
+
+```yaml
+tasks:
+- name: deploy-bb
+  kind: Operator
+  spec:
+    parameterFile: bb-params.yaml # optional, defines the parameter that will be set on the bb-instance  
```

Comment:
this seems overly complicated... it defines in the operator a static configuration by the operator dev for which the operator user would need to have deep insight into... it would be required of them to look at source to know what to do.

wouldn't a better approach be to scope parameters.... so `-p PASSWORD=foo` is what we have today...  perhaps that is just shorthand for `zookeeper:PASSWORD`.  If we need to allow for the same operator to be used for multiple tasks than perhaps we need a task id for the operator task

```yaml
tasks:
- name: deploy-bb
  id: foo
  kind: Operator
```
or we could use the name of the task (but I love the default to be the operator if we can / makes sense)

this would allow for `-p kafka:PASSWORD=foo -p zookeeper:PASSWORD=foo`

Reply:
I like the idea of prefixes. It might even be possible to adapt that to parameter files:
`--parameter-file zookeeper:zk-params.yaml --parameter-file kafka:kafka-params.yaml` and each of the parameters in that file would get the prefix automatically.

If we have multiple operators of the same type, we might need the full graph for the prefix: `-p AA:BB:PASSWORD=foo`

Reply:
> this seems overly complicated... it defines in the operator a static configuration by the operator dev for which the operator user would need to have deep insight into... it would be required of them to look at source to know what to do.

I think it's the opposite. If some app has a child RDBMS operator, it's an implementation detail of that app operator whether the RDBMS used is mysql or postgres. 
 It is up to its operator developer to:
- decide how much of the implementation to expose to end users, and in what way,
- document them, and how they affect nested operators, if any.

That way, the user of the top-level operator only has access to and needs to know about the parameters of the top-level operator. In the ideal case, 

IMHO forcing the user to know about how to configure a nested operator in detail, is creating complexity both on the user, and complicating matters for the top-level operator developer who will be unable to seamlessly switch to another operator once this detail has been leaked.

Also, I think @zen-dog already tried the prefix way, and that ended up with a [very complex implementation](https://github.com/kudobuilder/kudo/pull/1454#commits-pushed-1e200e4).

Reply:
As mentioned above `AA:BB:PASSWORD=foo` breaks the encapsulation and exposes the underlying implementation detail and _path to it_ to the end-user. This makes it much harder to treat e.g. underlying DB as an implementation detail. Also, the `EE` might decide to hard-code the password because it makes sense in that scenario and this detail might not be exposed at all the higher-level. 
In addition to the encapsulation argument, the proposed solution has the least amount of additional complexity in terms of implementation. I propose we start with it and reiterate if/when necessary.